pool/src/transaction_pool.rs

```diff
@@ -0,0 +1,175 @@
+
+/// A minimal (EXPERIMENTAL) transaction pool implementation
+pub struct TransactionPool<T> {
+	/// Pool Config
+	pub config: PoolConfig,
+
+	/// Our transaction pool.
+	pub txpool: Pool<T>,
+	/// Our Dandelion "stempool".
+	pub stempool: Pool<T>,
+
+	/// The blockchain
+	pub blockchain: Arc<T>,
+	/// The pool adapter
+	pub adapter: Arc<PoolAdapter>,
+}
+
+impl<T> TransactionPool<T>
+where
+	T: BlockChain,
+{
+	/// Create a new transaction pool
+	pub fn new(config: PoolConfig, chain: Arc<T>, adapter: Arc<PoolAdapter>) -> TransactionPool<T> {
+		TransactionPool {
+			config: config,
+			txpool: Pool::new(chain.clone(), format!("txpool")),
+			stempool: Pool::new(chain.clone(), format!("stempool")),
+			blockchain: chain,
+			adapter: adapter,
+		}
+	}
+
+	fn add_to_stempool(&mut self, entry: PoolEntry) -> Result<(), PoolError> {
+		// Add tx to stempool (passing in all txs from txpool to validate against).
+		self.stempool
+			.add_to_pool(entry.clone(), self.txpool.all_transactions())?;
+
+		// Note: we do not notify the adapter here,
+		// we let the dandelion monitor handle this.
+		Ok(())
+	}
+
+	fn add_to_txpool(&mut self, mut entry: PoolEntry) -> Result<(), PoolError> {
+		// First deaggregate the tx based on current txpool txs.
+		if entry.tx.kernels.len() > 1 {
+			let txs = self.txpool
+				.find_matching_transactions(entry.tx.kernels.clone());
+			if !txs.is_empty() {
+				entry.tx = transaction::deaggregate(entry.tx, txs)?;
+				entry.src.debug_name = "deagg".to_string();
+			}
+		}
+		self.txpool.add_to_pool(entry.clone(), vec![])?;
+
+		// We now need to reconcile the stempool based on the new state of the txpool.
+		// Some stempool txs may no longer be valid and we need to evict them.
+		let txpool_tx = self.txpool.aggregate_transaction()?;
+		self.stempool.reconcile(txpool_tx)?;
+
+		self.adapter.tx_accepted(&entry.tx);
+		Ok(())
+	}
+
+	/// Add the given tx to the pool, directing it to either the stempool or
+	/// txpool based on stem flag provided.
+	pub fn add_to_pool(
+		&mut self,
+		src: TxSource,
+		tx: Transaction,
+		stem: bool,
+	) -> Result<(), PoolError> {
+		// Do we have the capacity to accept this transaction?
+		self.is_acceptable(&tx)?;
+
+		// Make sure the transaction is valid before anything else.
+		tx.validate().map_err(|e| PoolError::InvalidTx(e))?;
+
+		// Check the tx lock_time is valid based on current chain state.
+		self.blockchain.verify_tx_lock_height(&tx)?;
+
+		// Check coinbase maturity before we go any further.
+		self.blockchain.verify_coinbase_maturity(&tx)?;
+
+		let entry = PoolEntry {
+			state: PoolEntryState::Fresh,
+			src,
+			tx_at: time::now_utc().to_timespec(),
+			tx: tx.clone(),
+		};
+
+		if stem {
+			self.add_to_stempool(entry)?;
+		} else {
+			self.add_to_txpool(entry)?;
+		}
+		Ok(())
+	}
+
+	/// Reconcile the transaction pool (both txpool and stempool) against the
+	/// provided block.
+	pub fn reconcile_block(&mut self, block: &Block) -> Result<(), PoolError> {
```

Comment:
If I'm not mistaken the previous version was returning evicted transactions, do we still have such functionality?

Reply:
We do not - I simplified this for now because we were not using the return value (the evicted txs) for anything.
It should be easy enough to add this back in if desired in the future.